crates/astria-bridge-withdrawer/src/withdrawer/ethereum/watcher.rs

```diff
@@ -0,0 +1,493 @@
+use std::sync::Arc;
+
+use astria_core::primitive::v1::{
+    asset,
+    asset::Denom,
+};
+use astria_eyre::{
+    eyre::{
+        eyre,
+        WrapErr as _,
+    },
+    Result,
+};
+use ethers::{
+    contract::LogMeta,
+    providers::{
+        Provider,
+        StreamExt as _,
+        Ws,
+    },
+    utils::hex,
+};
+use tokio::{
+    select,
+    sync::mpsc,
+};
+use tokio_util::sync::CancellationToken;
+use tracing::{
+    info,
+    warn,
+};
+
+use crate::withdrawer::{
+    batch::{
+        event_to_action,
+        Batch,
+        EventWithMetadata,
+        WithdrawalEvent,
+    },
+    ethereum::astria_withdrawer::AstriaWithdrawer,
+    state::State,
+};
+
+/// Watches for withdrawal events emitted by the `AstriaWithdrawer` contract.
+pub(crate) struct Watcher {
+    contract: AstriaWithdrawer<Provider<Ws>>,
+    event_tx: mpsc::Sender<(WithdrawalEvent, LogMeta)>,
+    batcher: Batcher,
+    state: Arc<State>,
+    shutdown_token: CancellationToken,
+}
+
+impl Watcher {
+    pub(crate) async fn new(
```

Comment:
Make this non-async and split out the construction code from the run-time initialization code.

The goal is to ensure that the service can shut down cleanly when getting a shutdown signal